quint/src/effects/base.ts

```diff
@@ -323,19 +324,21 @@ export function unifyEntities(va: Entity, vb: Entity): Either<ErrorTree, Substit
     return bindEntity(v1.name, v2).mapLeft(msg => buildErrorLeaf(location, msg))
   } else if (v2.kind === 'variable') {
     return bindEntity(v2.name, v1).mapLeft(msg => buildErrorLeaf(location, msg))
-  } else {
-    if (isEqual(v1, v2)) {
-      return right([])
-    }
-
-    if (v1.kind === 'union' && v2.kind === 'concrete') {
-      return mergeInMany(v1.entities.map(v => unifyEntities(v, v2)))
-        .map(subs => subs.flat())
-        .mapLeft(err => buildErrorTree(location, err))
-    }
-
-    if (v1.kind === 'concrete' && v2.kind === 'union') {
-      return unifyEntities(v2, v1)
+  } else if (isEqual(v1, v2)) {
+    return right([])
+  } else if (v1.kind === 'union' && v2.kind === 'concrete') {
+    return mergeInMany(v1.entities.map(v => unifyEntities(v, v2)))
+      .map(subs => subs.flat())
+      .mapLeft(err => buildErrorTree(location, err))
+  } else if (v1.kind === 'concrete' && v2.kind === 'union') {
+    return unifyEntities(v2, v1)
+  } else if (v1.kind === 'union' && v2.kind === 'union') {
+    if (intersectionWith(v1.entities, v2.entities, isEqual).length > 0) {
+      const s1 = { ...v1, entities: differenceWith(v1.entities, v2.entities, isEqual) }
+      const s2 = { ...v2, entities: differenceWith(v2.entities, v1.entities, isEqual) }
```

Comment:
Simply an optimization, but given that we already compute the intersection, we could use it to compute the difference and avoid (re-)iterating the entire arrays?

Reply:
Yep, I meant to do that after my initial validation and ended up forgetting about it. Thanks.